File: Cpp/670_Maximum_Swap.cpp
class Solution {
public:
    int maximumSwap(int num) {
        string s = to_string(num);
        unordered_map<int, int> index;
        
        for(int i = 0; i< s.size(); ++i) {
            index[s[i] - '0'] = i; 
        }
        
        for(int i = 0; i < s.size(); ++i) {
            for(int j = 9; j > s[i] - '0'; --j) {
                if(index[j] > i) {
                    swap(s[index[j]], s[i]);
                    return stoi(s);
                }
            }
        }
        
        return stoi(s);
    }
};

File: Kotlin/670_Maximum_Swap.kt
class Solution {
    fun maximumSwap(num: Int): Int {
        val s = num.toString().toCharArray()
        val index = HashMap<Char, Int>()
        
        for(i in s.indices) {
            index[s[i]] = i 
        }
        
        for(i in s.indices) {
            for(j in '9' downTo '0') {
                index[j]?.let{ index ->
                    if(s[i] < j && index > i) {
                        s[i] = s[index].also { s[index] = s[i] }
                        return String(s).toInt()   
                    }
                }
            }
        }
        
        return num
    }
}